File: apps/mobile/lib/src/api/dispute_api.dart
import 'dart:typed_data';
import 'authed_api.dart';

final _uuidPattern = RegExp(r'^[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[1-5][0-9a-fA-F]{3}-[89abAB][0-9a-fA-F]{3}-[0-9a-fA-F]{12}$');
const disputeCategories = ['non_delivery', 'item_condition', 'damage', 'pickup_issue', 'payment_issue', 'other'];

class MobileDisputeSummary {
  const MobileDisputeSummary({
    required this.id,
    required this.orderId,
    required this.category,
    required this.status,
    required this.outcome,
    required this.reason,
    required this.responseDueAt,
    required this.reviewDueAt,
    required this.appealDeadlineAt,
  });
  final String id;
  final String orderId;
  final String category;
  final String status;
  final String outcome;
  final String reason;
  final DateTime responseDueAt;
  final DateTime reviewDueAt;
  final DateTime? appealDeadlineAt;

  factory MobileDisputeSummary.fromJson(Map<String, dynamic> json) {
    final id = json['id'];
    final orderId = json['orderId'];
    final category = json['category'];
    final status = json['status'];
    final outcome = json['outcome'];
    final reason = json['reason'];
    final responseDue = DateTime.tryParse(json['responseDueAt']?.toString() ?? '');
    final reviewDue = DateTime.tryParse(json['reviewDueAt']?.toString() ?? '');
    if (id is! String || !_uuidPattern.hasMatch(id) || orderId is! String || !_uuidPattern.hasMatch(orderId) ||
        category is! String || !disputeCategories.contains(category) || status is! String || outcome is! String ||
        reason is! String || responseDue == null || reviewDue == null) {
      throw const FormatException('Invalid dispute summary');
    }
    final appealRaw = json['appealDeadlineAt'];
    final appealDeadline = appealRaw == null ? null : DateTime.tryParse(appealRaw.toString());
    if (appealRaw != null && appealDeadline == null) throw const FormatException('Invalid appeal deadline');
    return MobileDisputeSummary(
      id: id,
      orderId: orderId,
      category: category,
      status: status,
      outcome: outcome,
      reason: reason,
      responseDueAt: responseDue,
      reviewDueAt: reviewDue,
      appealDeadlineAt: appealDeadline,
    );
  }
}

class MobileDisputeDetail {
  const MobileDisputeDetail({required this.dispute, required this.responses, required this.evidence, required this.appealStatus, required this.paymentOperationStatus});
  final MobileDisputeSummary dispute;
  final List<String> responses;
  final List<MobileDisputeEvidence> evidence;
  final String? appealStatus;
  final String? paymentOperationStatus;

  factory MobileDisputeDetail.fromJson(Map<String, dynamic> json) {
    if (json['dispute'] is! Map || json['responses'] is! List || json['evidence'] is! List) {
      throw const FormatException('Invalid dispute detail');
    }
    final responses = (json['responses'] as List).map((entry) {
      if (entry is! Map || entry['responseText'] is! String) throw const FormatException('Invalid dispute response');
      return entry['responseText'] as String;
    }).toList(growable: false);
    final evidence = (json['evidence'] as List).map((entry) {
      if (entry is! Map) throw const FormatException('Invalid dispute evidence');
      return MobileDisputeEvidence.fromJson(Map<String, dynamic>.from(entry));
    }).toList(growable: false);
    final appeal = json['appeal'];
    final operation = json['paymentOperation'];
    return MobileDisputeDetail(
      dispute: MobileDisputeSummary.fromJson(Map<String, dynamic>.from(json['dispute'] as Map)),
      responses: responses,
      evidence: evidence,
      appealStatus: appeal is Map ? appeal['status']?.toString() : null,
      paymentOperationStatus: operation is Map ? operation['status']?.toString() : null,
    );
  }
}

class MobileDisputeEvidence {
  const MobileDisputeEvidence({required this.id, required this.type, required this.filename, required this.textValue, required this.downloadPath});
  final String id;
  final String type;
  final String? filename;
  final String? textValue;
  final String? downloadPath;

  factory MobileDisputeEvidence.fromJson(Map<String, dynamic> json) {
    final id = json['id'];
    final type = json['type'];
    if (id is! String || !_uuidPattern.hasMatch(id) || type is! String) throw const FormatException('Invalid dispute evidence');
    final path = json['downloadPath'];
    if (path != null && (path is! String || !path.startsWith('/v1/market/disputes/'))) throw const FormatException('Invalid private evidence path');
    return MobileDisputeEvidence(id: id, type: type, filename: json['filename'] as String?, textValue: json['textValue'] as String?, downloadPath: path as String?);
  }
}

abstract interface class DisputeGateway {
  Future<List<MobileDisputeSummary>> list(String accessToken);
  Future<MobileDisputeDetail> detail(String accessToken, String disputeId);
  Future<MobileDisputeSummary> open(String accessToken, {required String orderId, required String category, required String reason});
  Future<void> respond(String accessToken, {required String disputeId, required String responseText});
  Future<void> addTextEvidence(String accessToken, {required String disputeId, required String text});
  Future<void> uploadImageEvidence(String accessToken, {required String disputeId, required String filename, required String contentType, required Uint8List bytes});
  Future<void> appeal(String accessToken, {required String disputeId, required String reason});
}

class DisputeApi implements DisputeGateway {
  const DisputeApi({required AuthedApi authedApi}) : _api = authedApi;
  final AuthedApi _api;

  String _id(String value) {
    final result = value.trim();
    if (!_uuidPattern.hasMatch(result)) throw const FormatException('Invalid identifier');
    return result;
  }

  @override
  Future<List<MobileDisputeSummary>> list(String accessToken) async {
    final json = await _api.get('/v1/market/disputes?limit=100', accessToken);
    final rows = json['disputes'];
    if (rows is! List) throw const FormatException('Invalid dispute list');
    return rows.map((row) {
      if (row is! Map) throw const FormatException('Invalid dispute list row');
      return MobileDisputeSummary.fromJson(Map<String, dynamic>.from(row));
    }).toList(growable: false);
  }

  @override
  Future<MobileDisputeDetail> detail(String accessToken, String disputeId) async {
    return MobileDisputeDetail.fromJson(await _api.get('/v1/market/disputes/${_id(disputeId)}', accessToken));
  }

  @override
  Future<MobileDisputeSummary> open(String accessToken, {required String orderId, required String category, required String reason}) async {
    if (!disputeCategories.contains(category)) throw const FormatException('Invalid dispute category');
    final trimmed = reason.trim();
    if (trimmed.length < 20 || trimmed.length > 4000) throw const FormatException('Invalid dispute reason');
    final json = await _api.post('/v1/market/disputes', accessToken, {'orderId': _id(orderId), 'category': category, 'reason': trimmed});
    if (json['dispute'] is! Map) throw const FormatException('Invalid opened dispute');
    return MobileDisputeSummary.fromJson(Map<String, dynamic>.from(json['dispute'] as Map));
  }

  @override
  Future<void> respond(String accessToken, {required String disputeId, required String responseText}) async {
    final text = responseText.trim();
    if (text.length < 10 || text.length > 6000) throw const FormatException('Invalid dispute response');
    await _api.post('/v1/market/disputes/${_id(disputeId)}/responses', accessToken, {'responseText': text});
  }

  @override
  Future<void> addTextEvidence(String accessToken, {required String disputeId, required String text}) async {
    final value = text.trim();
    if (value.length < 3 || value.length > 10000) throw const FormatException('Invalid text evidence');
    await _api.post('/v1/market/disputes/${_id(disputeId)}/evidence/text', accessToken, {'evidenceType': 'participant_statement', 'text': value});
  }

  @override
  Future<void> uploadImageEvidence(String accessToken, {required String disputeId, required String filename, required String contentType, required Uint8List bytes}) async {
    if (!['image/jpeg', 'image/png', 'image/webp'].contains(contentType) || bytes.isEmpty || bytes.length > 10 * 1024 * 1024) {
      throw const FormatException('Invalid image evidence');
    }
    final query = Uri(queryParameters: {'evidenceType': 'participant_image', 'filename': filename.trim()}).query;
    await _api.postBinaryWithHeaders('/v1/market/disputes/${_id(disputeId)}/evidence/upload?$query', accessToken, bytes, contentType: contentType);
  }

  @override
  Future<void> appeal(String accessToken, {required String disputeId, required String reason}) async {
    final value = reason.trim();
    if (value.length < 20 || value.length > 4000) throw const FormatException('Invalid appeal reason');
    await _api.post('/v1/market/disputes/${_id(disputeId)}/appeal', accessToken, {'reason': value});
  }
}

File: apps/web/lib/dispute-api.ts
import { getAuthed, postAuthed, postAuthedBinary } from './authed-api';

const uuidPattern = /^[0-9a-f]{8}-[0-9a-f]{4}-[1-5][0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$/i;

export type DisputeCategory = 'non_delivery' | 'item_condition' | 'damage' | 'pickup_issue' | 'payment_issue' | 'other';
export type DisputeStatus = 'opened' | 'awaiting_buyer' | 'awaiting_seller' | 'under_review' | 'resolved' | 'closed';
export type DisputeOutcome = 'none' | 'buyer_refund' | 'seller_release' | 'partial_refund' | 'return_required' | 'compliance_escalation';

export interface DisputeSummary {
  id: string;
  orderId: string;
  paymentIntentId: string;
  openedByUserId: string;
  respondentUserId: string;
  openedByRole: 'buyer' | 'seller';
  category: DisputeCategory;
  status: DisputeStatus;
  outcome: DisputeOutcome;
  reason: string;
  summary: string | null;
  responseDueAt: string;
  reviewDueAt: string;
  resolutionPaymentOperationId: string | null;
  resolutionNotes: string | null;
  appealDeadlineAt: string | null;
  escalationLevel: number;
  escalationReason: string | null;
  openedAt: string;
  resolvedAt: string | null;
  lastActivityAt: string;
}

export interface DisputeDetail {
  dispute: DisputeSummary;
  responses: Array<{ id: string; submittedByUserId: string; responseText: string; submittedAt: string }>;
  evidence: Array<{
    id: string;
    submittedByUserId: string;
    type: string;
    filename: string | null;
    mimeType: string | null;
    sizeBytes: number | null;
    note: string | null;
    textValue: string | null;
    downloadPath: string | null;
    createdAt: string;
  }>;
  appeal: null | {
    id: string;
    openedByUserId: string;
    reason: string;
    status: string;
    decisionNotes: string | null;
    openedAt: string;
    decidedAt: string | null;
  };
  events: Array<{ id: string; actorUserId: string | null; type: string; details: Record<string, unknown>; occurredAt: string }>;
  paymentOperation: null | { id: string; kind: string; status: string; amount: string | number | null; currency_code: string; requested_at: string; completed_at: string | null };
}

function id(value: string, label: string): string {
  const normalized = value.trim();
  if (!uuidPattern.test(normalized)) throw new Error(`${label} must be a UUID`);
  return normalized;
}

export function listDisputes(): Promise<{ disputes: DisputeSummary[] }> {
  return getAuthed('/v1/market/disputes?limit=100');
}

export function readDispute(disputeId: string): Promise<DisputeDetail> {
  return getAuthed(`/v1/market/disputes/${id(disputeId, 'Dispute identifier')}`);
}

export function openDispute(input: {
  orderId: string;
  category: DisputeCategory;
  reason: string;
  summary?: string;
}): Promise<{ dispute: DisputeSummary }> {
  return postAuthed('/v1/market/disputes', {
    orderId: id(input.orderId, 'Order identifier'),
    category: input.category,
    reason: input.reason.trim(),
    ...(input.summary?.trim() ? { summary: input.summary.trim() } : {})
  });
}

export function respondToDispute(disputeId: string, responseText: string) {
  return postAuthed(`/v1/market/disputes/${id(disputeId, 'Dispute identifier')}/responses`, { responseText: responseText.trim() });
}

export function addTextDisputeEvidence(disputeId: string, input: { evidenceType: string; text: string; note?: string }) {
  return postAuthed(`/v1/market/disputes/${id(disputeId, 'Dispute identifier')}/evidence/text`, {
    evidenceType: input.evidenceType.trim(),
    text: input.text.trim(),
    ...(input.note?.trim() ? { note: input.note.trim() } : {})
  });
}

export function uploadDisputeEvidence(disputeId: string, input: { evidenceType: string; file: File; note?: string }) {
  const query = new URLSearchParams({ evidenceType: input.evidenceType.trim(), filename: input.file.name });
  if (input.note?.trim()) query.set('note', input.note.trim());
  return postAuthedBinary(
    `/v1/market/disputes/${id(disputeId, 'Dispute identifier')}/evidence/upload?${query.toString()}`,
    input.file
  );
}

export function appealDispute(disputeId: string, reason: string) {
  return postAuthed(`/v1/market/disputes/${id(disputeId, 'Dispute identifier')}/appeal`, { reason: reason.trim() });
}

export function protectedEvidenceHref(downloadPath: string): string {
  if (!downloadPath.startsWith('/v1/market/disputes/')) throw new Error('Invalid private evidence path');
  return `/api/authed${downloadPath}`;
}
